oscLib.js

```diff
@@ -1,82 +1,144 @@
-        console.error(`Getal is te klein of te groot\nHet getal mag niet kleiner zijn dan -2147483648 en groter dan 2147483647.\nhet getal dat je probeert te versturen is ${message}`)
+      else if (message < - 2147483648 || message > 2147483647) {
+        console.error(`Getal is te klein of te groot\nHet getal mag niet kleiner zijn dan -2147483648 en groter dan 2147483647.\nhet getal dat je probeert te versturen is ${message}`);
       }
       else {
-        let sendData = {"address":address,"message":message,"id":socket.io.engine.id};
+        const sendData = {"address":address,"message":message,"id":socket.io.engine.id};
         socket.emit('sendMessage',sendData);
       }
-    },
-    killClient: function() {
+    }
+  }
+  killClient() {
+    if (this.clientActive) {
       socket.emit('killClient');
+      this.clientActive = 0;
     }
-  } 
+  }
 }
 
-function makeNote(note=60,velo=0.5,dur=0.0) {
-  let env = new p5.Envelope(0.01,velo,dur,0.0);
-  let triOsc = new p5.Oscillator('triangle');
-  triOsc.freq(midiToFreq(note));
-  triOsc.start();
-  env.play(triOsc);
-}
 
-function Server() {
-  return {
-    startServer: function(port) {
-      let sendData = {"port":port,"id":socket.io.engine.id};
+/*
+  Class voor het aanmaken van een OSC-server. De volgende functies zijn beschikbaar:
+ * - startServer(poort) -> start een osc-server waarmee data ontvangen kan worden
+ * - getMessage(callback) -> ontvang data en stuurt via de callback twee waardes terug, adres en waarde (array van waarde als er meer dan 1 waarde is)
+ * - killServer() -> stopt de server (kan dit weg?)
+*/
+class Server {
+  constructor() {
+    this.serverActive = 0;
+    this.serverCounter = 0;
+  }
+
+  startServer(port) {
+    if (connected) {
+      const sendData = {"port":port,"id":socket.io.engine.id};
       socket.emit('startServer',sendData);
-    },
-    getMessage: function(callback) {
-      socket.on('getMessage',function(data){
-        let msg = data.msg;
-        if (data.msg.length == 1) {
-          msg = data.msg[0];
-        }
-        callback(data.add,msg);
-      });
-    },
-    killServer : function() {
+      this.serverActive = 1;
+    }
+    else {
+      this.serverCounter++;
+      if (this.serverCounter < 4) {
+        setTimeout(_=> {
+          this.startServer(port);
+        },100);        
+      }
+      else {
+        console.error(`Kan geen server opstarten...`);
+      }
+    }
+  }
+  getMessage(callback) {
+    socket.on('getMessage',data => {
+      let msg = data.msg;
+      if (data.msg.length == 1) {
+        msg = data.msg[0];
+      }
+      callback(data.add,msg);
+    });
+  }
+  killServer() {
+    if (this.serverActive) {
       socket.emit('killServer');
+      this.serverActive = 0;
     }
   }
 }
+
+
+function makeNoteSetup() {
+  env = new p5.Envelope();
+  triOsc = new p5.Oscillator('triangle');
+}
+
+function makeNote(note=60,velo=0.5,dur=0.0) {
+  env.setADSR(0.01,velo,0,velo,dur,velo);
```

Comment:
Zorgt dit er niet voor dat iedere keer de vorige noot afgekapt wordt en een nieuwe wordt gestart?

Reply:
[p5.PolySynth](https://p5js.org/reference/#/p5.PolySynth) lijkt polyfoon te werken, maar is standaard alleen beschikbaar met een sinewave.

Reply:
klopt, is inderdaad nog zo monofoon als een koe. Daar mag nog wel wat aan gedaan worden, ofwel via de PolySynth, of een array van oscillators die achterelkaar gebruikt worden. Dit lost eigenlijk alleen op dat er allemaal oscillators en envelopes worden gemaakt die (daar leek het in ieder geval op) niet opgeruimd worden.

Reply:
Ik heb een snelle implementatie gemaakt en getest van p5.PolySynth zoals Jochem suggereert. Is dit een beetje wat je bedoelt @JochemVanIterson ? Dit werkt prima met standaard 8 sinus oscillators, misschien leuk om daar wat opties voor de studenten aan toe te voegen. Ik weet niet zo goed hoe ik een suggestie op deze pull-request toevoeg dus ik doe het eventjes zo. Als ik dit beter anders kan doen (denk het wel) hoor ik dat graag.

variabelen vanaf regel 0:
`const socket = io.connect();
let connected = 0;
let polySynth;`

en vanaf 130:
```
function makeNoteSetup() {
  polySynth = new p5.PolySynth()
}

function makeNote(note=60,velo=0.5,dur=0.0) {
  userStartAudio();
  //console.log("Making note", int(note), "with vel", int(velo), "and duration", dur, "ms")
  polySynth.play(midiToFreq(note), int(velo), 0.0, dur / 1000.)
}`

Reply:
Je kan het snel testen door een sketch te maken met dit als inhoud:

```
function setup() {
  //plaats hier de code die maar één keer hoeft te worden uitgevoerd
  createCanvas(800,600);
  background(255);
}

function draw() {
  //plaats hier de code die continue herhaald moet worden.
}

function mousePressed() {
  console.log("mousePressed playing random note")
  makeNote(30 + random(40), 60 + random(60), 1000)
}
```

Reply:
Ik heb je implementatie toegevoegd aan oscLib en gepusht, dus zou nu in deze branch moeten staan.
Nu even testen :-)

Reply:
@n-marcus Lijkt prima te werken dit ja, [hier](https://github.com/processing/p5.js-sound/issues/47) staat nog e.e.a. over andere oscillatoren voor de polySynth, ik heb nu alleen geen zin om het uit te zoeken. Ik denk dat dit voor nu prima is qua noten-afspeel-functie. We hebben ook nog wel even de tijd om het verder te implementeren want we kunnen de studenten via update nieuwe versies laten downloaden. 
Als jij en @JochemVanIterson verder geen gekke dingen tegen zijn gekomen wil ik hem wel mergen voor donderdag, dan kunnen de studenten er mee aan de slag.

Reply:
Ik zie geen gekke dingen meer!